Add `ExponentialBackoffErrorHandler` to manage message visibility tim…

brun0-4ugusto opened the pull request:
## :loudspeaker: Type of change
<!--- Put an `x` in the boxes that apply -->
- [ ] Bugfix
- [X] New feature
- [X] Enhancement
- [ ] Refactoring


## :scroll: Description
New ExponentialBackoffErrorHandler to change message visibility timeout exponentially based on the number of receive Message Count

## :bulb: Motivation and Context
As mentioned in (#1313), this enhancement enables reporting an error back to AWS SQS by exponentially increasing the message's visibility timeout, allowing it to be retried.

See #1314 


## :green_heart: How did you test it?
Unit test, Integration test, manual tests with a sample application

## :pencil: Checklist
<!--- Put an `x` in the boxes that apply -->
- [X] I reviewed submitted code
- [X] I added tests to verify changes
- [X] I updated reference documentation to reflect the change
- [X] All tests passing
- [X] No breaking changes


## :crystal_ball: Next steps
- Update sample application

tomazfernandes commented on spring-cloud-aws-sqs/src/main/java/io/awspring/cloud/sqs/listener/errorhandler/ExponentialBackoffErrorHandler.java:
```diff
@@ -0,0 +1,146 @@
+		this(initialVisibilityTimeoutSeconds, multiplier, DEFAULT_MAX_VISIBILITY_TIMEOUT_SECONDS);
+	}
+
+	public ExponentialBackoffErrorHandler(int initialVisibilityTimeoutSeconds, double multiplier, int maxVisibilityTimeoutSeconds) {
```
How about instead of multiple constructors we create an `ExponentialBackoffErrorHandlerBuilder`?

tomazfernandes reviewed (commented):
Looking great @brun0-4ugusto and @rafaelcgpava!

I left a few of comments in the code, please let me know if they make sense.

We can add some debug or trace logs so we can follow whether visibility change was successful and to what value it was changed to for a given message id.

We also need to decide what happens if visibility change fails for a given message or set of messages - we should at least log the error clearly with context.

#### Abstractions

Regarding abstractions, I can see where we could benefit from code reuse, but since it's not too much logic perhaps some duplication is ok?

We can also have a `ErrorHandlerVisibilityHelper` class with some utility methods.

Please let me know your thoughts on all of this, thanks!

brun0-4ugusto commented on spring-cloud-aws-sqs/src/main/java/io/awspring/cloud/sqs/listener/errorhandler/ExponentialBackoffErrorHandler.java:
```diff
@@ -0,0 +1,146 @@
+	}
+
+	private void checkMultiplier(double multiplier) {
+		Assert.isTrue(multiplier >= 1, () -> "Invalid multiplier '" + multiplier + "'. Should be greater than " +
```
Hi @tomazfernandes! I agree with you that we should wait until someone requests this change. :smile_cat:

brun0-4ugusto commented:
Hello @tomazfernandes,

I hope you’re doing well. We have implemented the requested changes—they were excellent ideas! We also took the opportunity to incorporate the new ErrorHandlerVisibilityHelper into the ImmediateRetryAsyncErrorHandler class. Please let us know if you have any further suggestions for improvement.

Thank you again! :smile_cat:

tomazfernandes commented on spring-cloud-aws-sqs/src/main/java/io/awspring/cloud/sqs/listener/errorhandler/ErrorHandlerVisibilityHelper.java:
```diff
@@ -0,0 +1,61 @@
+ * @author Bruno Garcia
+ * @author Rafael Pavarini
+ */
+public class ErrorHandlerVisibilityHelper {
```
Can we add a unit test suite for this class?

tomazfernandes commented on spring-cloud-aws-sqs/src/main/java/io/awspring/cloud/sqs/listener/errorhandler/ExponentialBackoffErrorHandler.java:
```diff
@@ -0,0 +1,167 @@
+		int timeout = calculateTimeout(message);
+		Visibility visibility = ErrorHandlerVisibilityHelper.getVisibility(message);
+		logger.debug("Changing visibility timeout to {} - Message Id {}", timeout, message.getHeaders().getId());
+		return CompletableFutures.exceptionallyCompose(visibility.changeToAsync(timeout), throwable -> {
```
The CompletableFutures class was necessary when we were using Java 8.

On Java 17 we can use the equivalent `CompletableFuture` methods directly.

tomazfernandes reviewed (commented):
Thanks for the changes, it's looking good!

Left a couple of comments. Also, please add docs on the new error handler.

brun0-4ugusto commented:
Hi @tomazfernandes ,

Thank you for your feedback and comments. We have addressed the points you raised and added documentation for the new error handler. Please let us know if you have any further suggestions or questions.

tomazfernandes commented:
Thanks for the PR @brun0-4ugusto and @rafaelcgpava!

I have two more issues in mind regarding error handling:
- Create a `LinearBackOffErrorHandler`
- Create a `BackOffErrorHandlerFactory`, where we can set the parameters we want and it'll return the proper implementation.

What do you think? Would you like to contribute those as well?

Let me know your thoughts, thanks!

brun0-4ugusto commented:
Hi @tomazfernandes Thanks for the review!
We’d be happy to contribute those as well! 😺

brun0-4ugusto commented:
Hi @tomazfernandes, how are you?
Apologies for the delay. I’ve submitted a PR that provides a LinearBackOffErrorHandler and additional exponential jitter strategies. I’ll send another PR with the factory shortly. (#1422 )
